modules/core/upgrade.py

```diff
@@ -0,0 +1,551 @@
+CHECK_INTERVAL_SECS = 5
+
+# Max time value to allow for ensuring both AWS and services have same set after killing
+TIMEOUT_SECS = 300
+
+# Default number of unseals required
+VAULT_UNSEAL_COUNT = 3
+
+# Commands to ensure to be available before starting
+CMDS = [
+    'aws',
+    'consul',
+    'curl',
+    'nomad',
+    'jq',
+    'scp',
+    'ssh'
+]
+
+#
+# Assertion commands
+#
+
+
+def assert_n_quorum(n):
+    if n == 0:
+        sys.exit(
+            'This process can only be run when there are 3 or more servers to maintain quorum.')
+
+
+def assert_arg(flag, arg):
+    if not arg:
+        sys.exit('Flag "{}" must be set!'.format(flag))
+
+
+def assert_collection_len(collection, length):
+    if len(collection) != length:
+        sys.exit('Collection "{}" does not have expected length {}!'.format(
+            collection, length))
+
+
+def assert_file_exists(path):
+    if not os.path.exists(path):
+        sys.exit('"{}" does not exist!'.format(path))
+
+
+def assert_cmds_exist(cmds):
+    for cmd in cmds:
+        if not shutil.which(cmd):
+            sys.exit('Require command "{}"!'.format(cmd))
+
+
+def assert_same_instances(aws_instances, service_nodes):
+    if not aws_instances == service_nodes:
+        sys.exit('AWS set of instances: "{}" is different from current set of service nodes: "{}"! Aborting...'
+                 .format(aws_instances, service_nodes))
+
+
+#
+# Test
+#
+
+# Can be used to swap out kill_fn
+def fake_kill_fn(aws_instance):
+    print('Fake killing instance "{}"'.format(aws_instance))
+
+
+# Can be used to swap out check_fn
+def fake_check_fn(prev_instances, killing_idx):
+    print('Fake checking {} at killing index {}'.format(
+        prev_instances, killing_idx))
+    return True
+
+
+# Can be used to override the default get_new_instances_fn
+def fake_get_new_instances(prev_instances, curr_instances):
+    return list(curr_instances)[0]
+
+#
+# General
+#
+
+
+def unique(seq):
+    # Inefficient but works
+    # http://www.martinbroadhurst.com/removing-duplicates-from-a-list-while-preserving-order-in-python.html
+    seen = set()
+    return [x for x in seq if not (x in seen or seen.add(x))]
+
+
+def find_n_to_kill_in_quorum(instances):
+    # The number of instances left must be simple majority to maintain quorum
+    # https://www.consul.io/docs/internals/consensus.html#deployment-table
+    # Be careful with numeric operations in Python, since they are by default
+    # interpreted as floating points operations
+    return int(max(math.floor((len(instances) + 1) / 2) - 1, 0))
+
+
+def invoke_shell(cmd):
+    return subprocess.check_output(cmd, shell=True).decode('ascii').strip()
+
+
+def get_instance_ids_from_tag(tag_pattern):
+    try:
+        return set(invoke_shell("""
+        aws ec2 describe-instances --filter \
+            "Name=tag:Name,Values=*{}*" \
+            "Name=instance-state-name,Values=running" | \
+            jq --raw-output '.Reservations[].Instances[].InstanceId'
```

Comment:
Well if we allow `pip`, then our aim would be to completely remove the use of shell scripts, since I think there are packages to Consul, Nomad, Vault clients, as well as AWS. We can even use a better CLI args parser to possibly group the args based on the service types.

For a quick hack, let's go with this first.